specifications: unified provider (#3283)
snapcore/snapcraft: specifications/unified-provider.org

=snapcraft.internal.build_providers._base_provider.Provider= such that it would
support all current build providers (Host, Launchpad, LXD, Managed-Host,
Multipass) and potential future providers (Remote LXD, etc.).

* Provider Refactoring

** Current public interfaces
- ensure_provider()
- setup_provider()
- mount_project()
- expose_prime()
- execute_step()
- pack_project()
- clean(part_names)
- clean_project()
- shell()

** Proposed draft of public interfaces

#+begin_src python
class Provider(abc.ABC):
  def __init__(self, project: Project, **bp_kwargs) ->
  None:
    """Save instance attributes, take no actions."""

  def __enter__(self) -> Provider:
    """Start-of-session actions, not to include launching VMs/instances."""

  def __exit__(self) -> None:
    """End-of-session cleanup that may be configurable.

    Examples include:
    - scheduling stop for build-instance VM/container
    - removing temporary files or artifacts
    - purging ephemeral build-instance
    """

  def is_provider_ready(self) -> bool:
    """Check if provider is accessible.

      As applicable, validate that the provider is:
      - installed
      - configured
      - running
      - compatible

>       Used as a pre-check prior to initiating build. If False, indicates that

Author: ```suggestion
      Used as a pre-check prior to use. If False, indicates that
```

      attempt to make all necessary changes.

      :param interactive: Toggle for user input.

      :raises SnapcraftEnvironmentError: for failures setting up provider.
      """

    def execute_step(self, step: step.Step, part_name: Optional[List[str]]) -> None
      """Execute (up to, and including) specified step.

      :param step: Step to execute.
      :param part_names: Optional list of part names to execute step for.

      :raises SnapcraftProviderUnsupportedMethod: if provider does not support
          executing steps.
      """

    def expose_prime(self, prime_dir: pathlib.Path) -> None:
      """Make project's prime directory available to host's prime_dir.

      Implementations should use the most performant approach available
      for the given provider (i.e. bind mounting directory from container).

      :param prime_dir: Directory to mount/sync prime directory to on host.

      :raises SnapcraftProviderUnsupportedMethod: if provider does not support
          syncing/mounting project's prime directory.
      """

    def snap(self, output_dir: pathlib.Path) -> List[pathlib.Path]:

>       """Snap project, building as required.

Author: ```suggestion
      """Snap project, executing lifecycle steps as required.
```

- ensure_provider()
- setup_provider()
- mount_project()
- expose_prime()
- execute_step()
- pack_project()
- clean(part_names)
- clean_project()
- shell()

** Proposed draft of public interfaces

#+begin_src python
class Provider(abc.ABC):
  def __init__(self, project: Project, **bp_kwargs) ->
  None:
    """Save instance attributes, take no actions."""

  def __enter__(self) -> Provider:
    """Start-of-session actions, not to include launching VMs/instances."""

  def __exit__(self) -> None:
    """End-of-session cleanup that may be configurable.

    Examples include:
    - scheduling stop for build-instance VM/container
    - removing temporary files or artifacts
    - purging ephemeral build-instance
    """

  def is_provider_ready(self) -> bool:
    """Check if provider is accessible.

      As applicable, validate that the provider is:
      - installed
      - configured
      - running
      - compatible

      Used as a pre-check prior to initiating build. If False, indicates that
      additional installation, or configuration must be performed by user, which
      can be done with setup_provider().

      :returns: True if ready, False if further setup is required.

> 

Reviewer: ```suggestion
```

#+TITLE: Unified Provider
#+AUTHOR: Chris Patterson <chris.patterson@canonical.com>
#+DATE: [2020-09-11]

* Scope:

Snapcraft currently implements build providers for LXD and Multipass.  Host
builds are driven with independent logic, and Launchpad is indendently
implemented for remote-build.  With the re-introduction of remote LXD support,
it is desirable to refactor all providers into a consistent implementation that
has varied support for Snapcraft capabilities.

This document defines an approach to refactor build provider
=snapcraft.internal.build_providers._base_provider.Provider= such that it would
support all current build providers (Host, Launchpad, LXD, Managed-Host,
Multipass) and potential future providers (Remote LXD, etc.).

* Provider Refactoring

** Current Public Interfaces
- ensure_provider()
- setup_provider()
- mount_project()
- expose_prime()
- execute_step()
- pack_project()
- clean(part_names)
- clean_project()
- shell()

** Proposed Public Interfaces:

> |--------------------------------------------+---------------------------------------------------------------------------+-------------------------------------------------------|

Reviewer: How about making this class like (ABC) where methods are methods, descriptions and raises docstrings

Author: But org tables are so much fun... i'll do that.

| snap() -> List[pathlib.Path]               | snap project (building as required), writing snap(s) to host project dir. |                                                       |
| clean_parts(part_names: List[str]) -> None | clean project parts                                                       | NotImplementedError                                   |
| clean() -> None                            | clean all project artifacts, including build-instance, if applicable.     |                                                       |
| shell() -> None                            | launch interactive shell with build-instance, if applicable.              | NotImplementedError                                   |
|--------------------------------------------+---------------------------------------------------------------------------+-------------------------------------------------------|

*** Summary of changes:
- Rename =ensure_provider()= to =is_provider_ready()=, scoping to see if
  provider is ready without needing installation or further configuration.

>   Remove exceptions.  CLI will invoke setup_provider() if terminal is

Reviewer: I don't understand this statement, can you make it clearer?

Author: will do

Author: should be done

  interactive, allowing user to perform any install/configuration as necessary.

> - Remove =mount_project()=. Project path will be provided to __init__() and will

Reviewer: Can you detail what the scope of __init__ will be?

Author: I will put that in the class definition

  be verified/mounted as required.
- Replace =pack_project()= with =snap()=, adding the implication that build steps

>   are to be executed as required (and packing may not be done by Snapcraft).

Reviewer: (and packing may not be done by Snapcraft) -> please explain

Author: In that sense that Launchpad may pack the snap, Snapcraft just fetches it.  Will clarify.

- Rename =clean()= to =clean_parts()=, to improve clarity of scope.
- Rename =clean_project()= to =clean()=, to improve clarity of scope.

** Migrate current logic into new providers:
- HostProvider
- ManagedHostProvider, inherits much of HostProvider.
- LaunchpadProvider
- RemoteLXDProvider, inherits much of LXDProvider, overrides the (now internal)
  equivalent of mount_project() behavior with a rsync from host to remote.

* CLI Handling

** Provider NotImplementedError

Certain commands will not supported by all providers. E.g.

> - =snapcraft try= is not compatible with Launchpad, Remote LXD, Host

Reviewer: it should be a noop for Host, in the future, we should call the snap command that does the try for you

Reviewer: maybe that should be part of this work :-) (as in defined here)

Author: Good idea!

- =snapcraft <step>= is not compatible with Lauchpad, Remote LXD

> - =snapcraft clean <parts>= is not compatible with Launchpad, Remote LXD

Reviewer: these latter two would be supported with remote-lxd in the future, we will just progressively implement. Care to describe that so the full vision is presented?

There is also no comment about --shell, --shell-after and --debug

Author: Ah yes, good points.

I'll write it such that Remote LXD will support <step> and <parts>.  I'm actually not sure why I wrote it this way :thinking: 

Will add that Launchpad will not support --shell, --shell-after and --debug.

In these cases, the applicable CLI methods will catch the NotImplementedError
and inform the user that this command is not available with the current
configured provider, e.g.:

> =Sorry, '<command>' is not available for the currently configured provider

Reviewer: Don't say sorry :-)

Author: Sorry!

'<provider>'.=

For =snapcraft clean <parts>=:

> =Sorry, specifying parts '<parts>' for 'snapcraft clean' is not supported by the

Reviewer: ditto

Author: will fix

#+AUTHOR: Chris Patterson <chris.patterson@canonical.com>
#+DATE: [2020-09-11]

* Summary

** What are we trying to do

Unify all build providers with a set of interfaces that can support
all providers, each with their own limitations and capabilities.

> This will allow us to more easily introduce new providers, such
> as Remote LXD, and improve consistency for existing ones.

Reviewer: ```suggestion
This will allow us to more easily introduce new providers and improve
consistency for existing ones.
```

- clean_project()
- shell()

** Proposed draft of public interfaces

#+begin_src python

> class ProviderV2:

Reviewer: Is this abstract?

> class ProviderV2:

Reviewer: Just call it `Provider`, we won't have two versions of this living for long as we do for plugins and it is all internal anyways

  def __init__(self, project_dir: pathlib.Path, project: Project, environment_manager_db: tinydb.TinyDB, ...) ->
  None:
    """Save instance attributes, take no actions."""

>     ...

Reviewer: ```suggestion
```

  def __enter__(self) -> Provider:
    """Start-of-session actions, not to include launching VMs/instances.

    Examples include:
    - querying datastore for relevant properties/configuration.
    """

>     ...
>     return self

Reviewer: The signature explains return self
```suggestion
```

  def __exit__(self) -> None:
    """End-of-session cleanup that may be configurable.

    Includes closing datastore handle and any provider-specific
    actions that are configured to be taken.

    Examples include:
    - scheduling stop for build-instance VM/container
    - removing temporary files or artifacts
    - purging build-instance
    """

  def is_provider_ready(self) -> bool:
    """Check if provider is accessible.

      As applicable, validate that the provider is:
      - installed
      - configured
      - running
      - compatible

>       Used as a pre-check prior to initiating build. If False, indicates that

Reviewer: build is a loaded word and can mean many things, mind picking something else?

      additional installation, or configuration must be performed by user, which
      can be done with setup_provider().

      :returns: True if ready, False if further setup is required.

      """

> 
>       ...

Reviewer: ```suggestion
```

    def setup_provider(self, interactive: bool) -> None:
      """Install/configure as necessary.

      If interactive, prompt user for all necessary changes. If non-interactive,
      attempt to make all necessary changes.

>       :param interactive: Toggle for user input.

Reviewer: How is user input handled?

Author: Need to add echoer

      :raises ProviderSetupError: if provider has dependencies that need to be
      installed, or configuration that must be changed.
      """

> 
>       ...

Reviewer: ```suggestion
```

    def execute_step(self, step: step.Step, part_name: Optional[List[str]]) -> None
      """Execute (up to, and including) specified step.

      :param step: Step to execute.
      :param part_names: Optional list of part names to execute step for.

      :raises SnapcraftProviderUnsupportedMethod: if provider does not support
      executing steps.
      """

> 
>       ...

Reviewer: ```suggestion
```

    def expose_prime(self, prime_dir: pathlib.Path) -> None:
      """Make project's prime directory available to host's prime_dir.

      Implementations should use the most performant approach available
      for the given provider (i.e. bind mounting directory from container).

      :param prime_dir: Directory to mount/sync prime directory to on host.

      :raises SnapcraftProviderUnsupportedMethod: if provider does not support
      syncing/mounting project's prime directory.
      """

    def snap(self, output_dir: pathlib.Path) -> List[pathlib.Path]:

>       """Snap project, building as required.

Reviewer: different word for building - > running through lifecycle steps

Reviewer: or "executing lifecycle steps" which matches the previous docstring

      Write output snaps to host project directory.

      :param output_dir: Directory to write snaps to.

      :returns: Path to snap(s) created from build.
      """

> 
>       ...

Reviewer: ```suggestion
```

     def clean_parts(self, part_names: List[str]) -> None:
      """Clean specified parts.

      :param part_names: List of parts to clean.

      :raises SnapcraftProviderUnsupportedMethod: if provider does not support

>       executing steps.

Reviewer: We usually anchor text after the colon

      """

> 
>       ...

Reviewer: ```suggestion
```

    def clean(self) -> None:
      """Clean all artifacts of project and build environment.

      Purges all artifacts from using the provider to build the
      project.  This includes build-instances (containers/VMs) and
      associated metadata and records.

      This does not include any artifacts that have resulted from
      a call to snap(), i.e. snap files or build logs.
      """

> 
>       ...

Reviewer: ```suggestion
```

    def shell(self, step: Optional[step.Step], part_name: Optional[str]) -> None:
      """Launch an interactive shell to build-instance.

      If available, load environment used for the given step and part_name.
      Step and part_name must both be supplied for environment to be loaded,
      otherwise both are ignored.

      Note this method does not imply execute_step(), it is up to the caller
      to launch the shell at the appropriate time.

      :param step: step to load environment for.
      :param part_name: part_name to load environment for.  If step is specified,

>       but not a part_name.

Reviewer: anchor

      :raises SnapcraftProviderUnsupportedMethod: if provider does not support

>       providing a shell.

Reviewer: anchor

      """

> 
>       ...

Reviewer: ```suggestion
```

#+end_src

*** Summary of changes against current provider:
- =is_provider_ready()= replaces classmethod =ensure_provider()=.
- =setup_provider()= becomes an instance method, supporting non-interactive
  configuration, if possible.
- Removes =mount_project()= interface. Provider will mount project as required.
- Replace =pack_project()= with =snap()=, adding the implication that build steps
  are to be executed as required, removing the implication that Snapcraft itself
  invokes the packing (Snapcraft may just fetch the output of the provider).
- Rename =clean()= to =clean_parts()=, to improve clarity of scope.
- Rename =clean_project()= to =clean()=, to improve clarity of scope.

> - Context manager =__enter__()= and =__exit__()= are currently used to define

Reviewer: This feel unbalanced, if all queries are related to the tinydb instance, why not query using tinydb directly (datastore instance) and then pass the relevant parameters into the provider?

Author: My original thought wass that the provider would know how to interpret the record(s) relevant to the provider.  But I can imagine another set of interfaces could manage that and pass it off as provider-specific kwargs.  That's probably preferable.

Author: I removed the datastore reference, I agree that shouldn't be there.

WRT 'imbalanced', are you referring to the shutdown of instance?  The other option was add a `close()` method and have the caller take care of it (probably with a `finally`).  I think I'd rather do that, but I mostly went this route because I couldn't come up with a good method name. :smirk:  Both approaches accomplish the same thing.  I'll add `close()` if you're good with it.

Reviewer: Why not just "stop"? Opposite to "start"

Author: `start()` is not an interface..?

Author: Will come back with draft example

  the lifetime start/stop status of containers. =__enter__()= will no longer
  start an instance, but =_exit()= will continue to stop the instance, if
  configured (and applicable) to do so.

* CLI Refactoring

** SnapcraftProviderUnsupportedMethodError

Certain commands will not supported by all providers. E.g.
- =snapcraft <step>= is not compatible with Lauchpad.
- =snapcraft clean <parts>= is not compatible with Launchpad

In these cases, the applicable CLI methods will catch the
=SnapcraftProviderUnsupportedMethodError= and inform the user that this command
is not available with the current configured provider.

*** Lifecycle command =snapcraft <step> ...=

If attempting to run a lifecycle command (other than =snap=) and
the provider does not support it, fail with the following error:
='<command>' is not available for the currently configured provider
'<provider>'.=

> For =--debug=, =--shell=, and =--shell-after= options:

Reviewer: should these be enforced by Provider sanity checks and not the cli implementation? It is not clarified here.

Author: I was planning on cli and having the cli handle the exception that it's not supported.  However I just realized some operations may run before it hits the exception. Perhaps it'd be best to have the providers advertise the features they support, or add to provider sanity checks.  I'll consider it further

Reviewer: I would prefer that the "capabilties" be described by the provider and the cli to raise an exception

Author: :+1:
